### PR TITLE
Mppa device

### PR DESCRIPTION
**Pull request by nicoTolly**
THIS BRANCH IS NOT SUPPOSED TO BE MERGED AS IS.

This is a work in progress for mppa port. Currently, telamon generates monothread, non-vectorized kernels that compile, are spawned correctly on telajax, and, for most but a few of them, output the correct result. For some of the kernel though, the output is incorrect. The problem seems not ot be a matter of float computation accuracy. Indeed, errors are still occuring with precision margin up to 10e-1. Another problem is that every kernel hangs on a call to pthread_barrier_wait which seems to be correct - waiting or one thread when no pthread_create has been called. Julien Hascoet is apparently the right guy to ask.

Namer and Namemap both have been seriously rewritten. We need to see how much of that can/should be keeped in final version. Most changes are intended to expose just name_map in printer and to call namer only through name_map. This change turned out to be quite unnecessary, but could still be benefic nonetheless (the dichotomy name_map/namer can be confusing)

### CODE REVIEW

**Comment by ulysseB** [comment]
Awesome work. If think we can do a lot better than the current state (in master) of the namer. Maybe we can use the transfer to using values to do things differently. I you be glad to hear any insights you have on that subject.

I only gave a quick look to your PR (will look in details over the week-end). I noticed you added a git submodule for Telajax. I would prefer if you create a separate repository with a rust library that exposes telajax bindings (telajax-sys) and use Cargo.toml to depend on it. I think it makes things easier to manage.

File: kernels/src/kernel.rs [hunk 1]
```diff
@@ -72,6 +72,7 @@ pub trait Kernel<'a>: Sized {
                     "evaluation failed for kernel {}, with actions {:?}",
                     Self::name(), leaf.actions);
                 if let Err(err) = kernel.check_result(&expected_output, context) {
+                    std::fs::write("actions_dump.txt", format!("{:?}", leaf.actions)).expect("Could not write actions dump");
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply kernels/src/kernel.rs]: Logging would be better, but I wanted a quick fix. What I really need is to get the changes made by Basile on the logger, which I will do very soon.

File: src/codegen/namer.rs [hunk 2]
```diff
@@ -17,8 +17,29 @@ pub enum Value<'a> {
     Operand(&'a ir::Operand<'a>),
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
+pub enum VarType {
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/namer.rs]: This is a bit more restrictive than ir::Type in the sense that we don't have a Void type (these are only types that can be used for a concrete variable) and that Ptr does not take a memory id. I should check why I need a different Type. Basile suggests a name 'Kind' that is apparently more adequate with respect to type theory

File: src/codegen/namer.rs [hunk 3]
```diff
@@ -27,13 +48,15 @@ pub trait Namer {
     fn name_float(&self, &Ratio<BigInt>, u16) -> String;
     /// Provides a name for an integer constant.
     fn name_int(&self, &BigInt, u16) -> String;
+    /// Returns an iterator on tuples (type t, number of declared variables of type t)
+    fn get_declared_variables(&self) -> Vec<(VarType, usize)>;
```
[the remark this answers is not in the document]
Comment (nicoTolly) [reply src/codegen/namer.rs]: We need it in var_decls (in both x86/printer and mppa/printer) which used to access Namer directly in order to do that and declare every variables. Now we do that through NameMap and this function

File: src/codegen/namer.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ulysseB) [reply src/codegen/namer.rs]: Ok. Note that the Void type was removed a while ago.
